### PR TITLE
Add ability to cache package payloads synchronously

### LINKED ISSUE #1379: Should be possible to configure that a resolve blocks until caching completes (and use the cache for all items)
Subject says most of it.

Ideally this would be implemented as a rezconfig flag, and perhaps rez-env or similar would have a flag to ignore the caching step if a user merely wishes to check the resolve. `--solve-only`?

I imagine that this sort of workflow would be a requirement of s3-bucket-style repositories anyway, so we ought to support this in general.

**Motivation**
Some users may wish to enforce usage of only cached packages, and simply wait for caching to complete.

### PR DESCRIPTION
**Pull request by isohedronpipeline**
Fixes #1379

Carrying on the work from https://github.com/AcademySoftwareFoundation/rez/pull/1452

This adds support for synchronous/blocking package caching during the resolve process.

### CODE REVIEW

File: src/rez/cli/env.py
Comment (JeanChristopheMorinPerso) [inline src/rez/cli/env.py:122]:
```suggestion
    parser.add_argument(
        "--pkg-cache-mode", choices=["sync", "async"],
        help="Package caching mode")
```
I feel like this would cover more cases. With --pkg-cache-sync, if a user sets package_cache_async to false in their config, it would be impossible to undo from the CLI (ignoring setting an env var).

**Review by JeanChristopheMorinPerso — COMMENTED** [review]
Thanks for creating this PR @isohedronpipeline! I left some comments, please let me know if you have questions.

File: src/rez/tests/test_package_cache.py
Comment (isohedronpipeline) [reply src/rez/tests/test_package_cache.py:141]: I like proving that the config set it to True

File: src/rez/tests/test_package_cache.py
Comment (isohedronpipeline) [reply src/rez/tests/test_package_cache.py]: How about `Packages were expected to be cached, but were not.`?

File: src/rez/resolved_context.py
Comment (isohedronpipeline) [reply src/rez/resolved_context.py:255]: This causes the member attribute to be set twice.  We should move the next line into an `else` block if that is preferred.  However, the one I posted matches the pattern used by the block above:
```
        if package_caching is None:
            if building:
                package_caching = config.package_cache_during_build
            else:
                package_caching = True
        self.package_caching = package_caching
```

File: src/rez/package_cache.py
Comment (isohedronpipeline) [reply src/rez/package_cache.py]: :+1:
https://stackoverflow.com/a/30984882 seems to back you up :-)

File: src/rez/package_cache.py
Comment (isohedronpipeline) [reply src/rez/package_cache.py:379]: :+1:

File: src/rez/cli/env.py
Comment (isohedronpipeline) [reply src/rez/cli/env.py]: :+1:

File: src/rez/cli/env.py
Comment (JeanChristopheMorinPerso) [inline src/rez/cli/env.py:121]: This sentence is really hard to read. I'm not even sure it's a valid English sentence.

File: src/rez/package_cache.py
Comment (JeanChristopheMorinPerso) [inline src/rez/package_cache.py:375]:
```suggestion
        .. deprecated:: 3.1.0
           Use :method:`add_variants` instead.
```

File: src/rez/rezconfig.py
Comment (JeanChristopheMorinPerso) [inline src/rez/rezconfig.py:284]:
```suggestion
# .. versionadded:: 3.1.0
```
My bad on the version.

File: src/rez/cli/env.py
Comment (isohedronpipeline) [reply src/rez/cli/env.py]: Ha! Okay, try now.

File: src/rez/package_cache.py
Comment (JeanChristopheMorinPerso) [inline src/rez/package_cache.py:464]: Don't we want to keep `--daemon` when async?

Comment (isohedronpipeline) [reply src/rez/package_cache.py:464]: This list is identical to the `args` variable defined above, so I just reused `args`.

Comment (isohedronpipeline) [reply src/rez/package_cache.py:464]: `args = [exe, "--daemon", self.path]` on line 463.

Comment (JeanChristopheMorinPerso) [reply src/rez/package_cache.py:464]: Oh Isee, my bad. Should we remove `--daemon` for sync? Not sure if it really makes sense to launch it as a daemon process when using sync mode.

Comment (JeanChristopheMorinPerso) [reply src/rez/package_cache.py:464]: Actually, do we even need to start a subprocess when we synchronously cache? Also, what will happen to pending variants (see line 425)? If I understand the code, an in-flight variant would be skipped, which would result in a potentially broken environment...

Comment (isohedronpipeline) [reply src/rez/package_cache.py:464]: We probably don't need it no.  I was more or less copying the logic from the original PR for that.

There also seems to be a fair amount of logic in the CLI entry point for gathering things for the cache, so I'm not sure if we can really run it without the CLI.  for instance, look at `rez/cli/pkg-cache` line 154.  That seems to be the primary business logic of caching without the daemon.  If we were to remove the subprocess, we should move that whole block into the PackageCache class, which feels maybe beyond the scope of what we were trying to do, but maybe not.  Let me know if you'd like to forge ahead with that strategy.

**Comment by JeanChristopheMorinPerso** [comment]
I took a deeper look, and as things are in this PR, it's doesn't work as you would expect. If you were to `rez-env maya --pkg-cache-mode async` and then `rez-env maya --pkg-cache-mode sync` and initial cache (when we ran with async) takes long enough, the second rez-env (in sync mode) will skip the maya package. See https://github.com/AcademySoftwareFoundation/rez/blob/main/src/rez/package_cache.py#L210-L219. Basically, the package cache will skip any variant that is marked as in-progress.

Additionally, I still have to verify, but I think that we should really avoid using `--daemon` when caching synchronously. If we use `--daemon`, it would mean that cancelling `rez-env --pkg-cache-mode sync` would not cancel the caching process. After all, keeping the process alive is the the reason to have a daemon mode in the first place. Or I'd go even further, we should not use a subprocess.

**Comment by isohedronpipeline** [comment]
> I took a deeper look, and as things are in this PR, it's doesn't work as you would expect. If you were to `rez-env maya --pkg-cache-mode async` and then `rez-env maya --pkg-cache-mode sync` and initial cache (when we ran with async) takes long enough, the second rez-env (in sync mode) will skip the maya package. See https://github.com/AcademySoftwareFoundation/rez/blob/main/src/rez/package_cache.py#L210-L219. Basically, the package cache will skip any variant that is marked as in-progress. This would result in a non-localized environment.

I think it would be preferable to call `add_variant(force=True)` rather than poll until the `.copying` sentinel file has disappeared, but I'm not sure what to do about the file lock.  We really don't want to be copying the same path at the same time.  Perhaps we should modify this logic to copy to an intermediate location that does an atomic move operation once the copy has completed to avoid errors with multiple cache operations running concurrently.  I'm not sure there is a *great* solution for mixing sync and async caching operations for the same files at the same time.

> 
> Additionally, I still have to verify, but I think that we should really avoid using `--daemon` when caching synchronously. If we use `--daemon`, it would mean that cancelling `rez-env --pkg-cache-mode sync` would not cancel the caching process. After all, keeping the process alive is the the reason to have a daemon mode in the first place. Or I'd go even further, we should not use a subprocess.
Yeah, I'm fine with that.  I'll move some of the logic from the cli entry point into `package_cache` later today or tomorrow.

> 
> We should also consider adding a progress bar or some logs when using the sync mode. I anticipate that users will complain if `rez-env maya houdini` blocks for multiple minutes without clearly saying what's going on.
We can have something print out in the `_while_copying()` function that is threaded.  Since we're using `shutil.copytree` I don't think we're going to have much by the way of stats about total size/amount transferred/estimated time to completion without hammering the disk, which I'd like to avoid.

**Comment by JeanChristopheMorinPerso** [comment]
>I think it would be preferable to call add_variant(force=True) rather than poll until the .copying sentinel file has disappeared, but I'm not sure what to do about the file lock. We really don't want to be copying the same path at the same time. Perhaps we should modify this logic to copy to an intermediate location that does an atomic move operation once the copy has completed to avoid errors with multiple cache operations running concurrently. I'm not sure there is a great solution for mixing sync and async caching operations for the same files at the same time.

I'm not too sure why we would use `force=True` or why you don't want to poll. Can you explain your thought a bit more please? Like you say, we really don't want to copy twice or copy the same package multiple times in parallel. The code goes to great length to avoid this.

**Comment by isohedronpipeline** [comment]
> I'm not too sure why we would use `force=True` or why you don't want to poll. Can you explain your thought a bit more please? Like you say, we really don't want to copy twice or copy the same package multiple times in parallel. The code goes to great length to avoid this.

You could be right.  I am worried that a previous sync process could have failed, leaving behind the .copying file, but no further updates to it will have been made.  I would imagine that one use case for doing a package cache syncronously would be to force all of the caching operations to happen *right now* as possibly a way to restart a locked or frozen cache operation.

I see that the copying file is updated every .2 seconds though, so we should be able to check that and continue to wait while that file is being updated.

My hesitation to hit the disk repeatedly is maybe misplaced since the cache is on the local disk.  In studio environments, I hesitate to hit network drives with something that could be run many times in parallel.  Of course you're right and starting multiple large file transfers instead is madness.

**Comment by JeanChristopheMorinPerso** [comment]
The package cache class already knows how to handle stalled packages. It should be able to handle the case where a variant is left in progress due to an error or whatever else and restart it.

As for polling, I agree that polling a shared file system is not wanted. Though, caching is usually done on local disks. We could maybe use `inotify` to optimize ans simplify the process.

File: src/rez/package_cache.py
Comment (JeanChristopheMorinPerso) [inline src/rez/package_cache.py:244]: I think we should use the qualified name which will include the variant id.
```suggestion
                logger.warning(f"Not caching {variant.qualified_name}. "
```
The same applies to all the other places where `{variant.name}-{variant.version}` is use.

File: src/rez/package_cache.py
Comment (JeanChristopheMorinPerso) [inline src/rez/package_cache.py:962]: Can we use `rez.vendor.progress.spinner` instead? We also have a small wrapper in rez.util.ProgressBar` that we use in rez-search (https://github.com/AcademySoftwareFoundation/rez/blob/main/src/rez/package_search.py#L69).

**Review by JeanChristopheMorinPerso — COMMENTED** [review]
I did a first read of your recent changes, but I have not yet reviewed them thoroughly. Hopefully that will allow me to think about it a little bit more.

File: src/rez/package_cache.py
Comment (isohedronpipeline) [reply src/rez/package_cache.py]: Yep, updated.
It's not really a progress bar since it's just waiting for the status to change.  I suppose we could monitor the state of each copying package, but I think a spinner is generally okay for this.

File: src/rez/package_cache.py
Comment (isohedronpipeline) [reply src/rez/package_cache.py]: :+1:

**Comment by chadrik** [comment]
Hi @JeanChristopheMorinPerso  I think all the outstanding issues have been addressed.  Any other thoughts?

File: src/rez/package_cache.py
Comment (JeanChristopheMorinPerso) [inline src/rez/package_cache.py:255]: I think this else is not needed and is redundant.

File: src/rez/package_cache.py
Comment (JeanChristopheMorinPerso) [inline src/rez/package_cache.py:267]:
```suggestion
                    if status == self.VARIANT_FOUND:
                        # We have resolved into a satisfactory state
                        logger.info(f"{variant.qualified_name} "
                                    f"{self.STATUS_DESCRIPTIONS[status]}")
                    else:
                        logger.warning(f"{variant.qualified_name} "
                                       f"{self.STATUS_DESCRIPTIONS[status]}")
```
If I'm not mistaken, this would have the same result.

File: src/rez/package_cache.py
Comment (JeanChristopheMorinPerso) [inline src/rez/package_cache.py:417]: TODO for maintainers: Update with appropriate version number before merging.

File: src/rez/package_cache.py
Comment (JeanChristopheMorinPerso) [inline src/rez/package_cache.py:427]:
```suggestion
        """Add the given variants to the package payload cache.
        """
```

File: src/rez/package_cache.py
Comment (JeanChristopheMorinPerso) [inline src/rez/package_cache.py:647]: I feel like this method should be private unless there is a god reason not to. I also think `wait_for_copying` should be True by default.

File: src/rez/package_cache.py
Comment (JeanChristopheMorinPerso) [inline src/rez/package_cache.py:132]: I think `wait_for_copying` should be True by default.

File: src/rez/rezconfig.py
Comment (JeanChristopheMorinPerso) [inline src/rez/rezconfig.py:284]: TODO for maintainers: Update with appropriate version number before merging.

File: src/rez/package_cache.py
Comment (JeanChristopheMorinPerso) [inline src/rez/package_cache.py:250]: If I'm not mistaken, the `progress` package takes care of disabling spinners if there is no TTY attached the the process. And I also think the output goes to stderr by default. Could you confirm both please? I don't want this code to spam logs in CI or in a render farm and I also don't want to see the spinner in stdout since that would affect the pipebility (if that's a word...) of rez.

File: src/rez/package_cache.py
Comment (JeanChristopheMorinPerso) [inline src/rez/package_cache.py:259]: I don't think it's something we want to tackle here, but it could be a surprising behavior that nothing is done to force a copy when a variant is stalled.

**Review by JeanChristopheMorinPerso — COMMENTED** [review]
Thanks @isohedronpipeline. I think we are almost there. I left some small comments that should hopefully be quick to address.

It would be great if you could also update https://rez.readthedocs.io/en/stable/caching.html#package-caching. The behavior of the async caching is well documented. I think the behavior of the sync cache should also be documented in a similar way.

**Comment by ameliendeshams** [comment]
Hi @isohedronpipeline,

I'm delighted to see this pull request! We've been waiting for this feature, and it's great to see it implemented.

At Fortiche, we've tested this branch, and it works exactly as expected. The synchronous package caching is a game-changer for our studio environments, where we often need to cache packages quickly and reliably.

We're also happy to see that you've addressed the concerns about progress and logging in sync mode. It's essential to provide users with clear feedback when running long-running commands.

Thanks again for your work on this feature! We're excited to see it merged and start using it in our production environments.

File: src/rez/package_cache.py
Comment (isohedronpipeline) [reply src/rez/package_cache.py]: No, I think the description of a stalled cache in the documentation makes it clear that it needs to be cleaned by hand in order to cache again in the future.  I don't think we want to change that behavior here.

File: src/rez/package_cache.py
Comment (isohedronpipeline) [reply src/rez/package_cache.py:132]: That would change the default behavior from async to sync, which I don't think we want to do, do we?

File: src/rez/package_cache.py
Comment (isohedronpipeline) [reply src/rez/package_cache.py:641]: I'll make it private, but I'd rather not change the default beahvior of the caching from async to sync here.

Comment (isohedronpipeline) [reply src/rez/package_cache.py:641]: Sorry, this one can be true, you're right.

File: src/rez/package_cache.py
Comment (isohedronpipeline) [reply src/rez/package_cache.py:250]: Yeah, that appears to be the case.

File: src/rez/package_cache.py
Comment (JeanChristopheMorinPerso) [reply src/rez/package_cache.py:255]: This comment is till valid and should be addressed. https://docs.python.org/3/reference/compound_stmts.html#the-while-statement

>This repeatedly tests the expression and, if it is true, executes the first suite; if the expression is false (which may be the first time it is tested) the suite of the else clause, if present, is executed and the loop terminates.
>
>A [break](https://docs.python.org/3/reference/simple_stmts.html#break) statement executed in the first suite terminates the loop without executing the else clause’s suite. A [continue](https://docs.python.org/3/reference/simple_stmts.html#continue) statement executed in the first suite skips the rest of the suite and goes back to testing the expression.

**Review by JeanChristopheMorinPerso — CHANGES_REQUESTED** [review]
Almost there!

**Review by JeanChristopheMorinPerso — APPROVED** [review]
Thanks a lot @isohedronpipeline and @ttrently!

**Comment by JeanChristopheMorinPerso** [comment]
I just realized that we missed adding a CLI flag to the `rez-pkg-cache` command...

**Pull request merged**
